File: Scripts/R/ForSPART/1-simulate_LUT.R
# ==============================================================================
# ForSPART course pipeline, step 1/3: simulate a LUT with ToolsRTM::SPART()
# (Soil-Plant-Atmosphere Radiative Transfer -- couples fourSAIL canopy +
# BSM/MARMIT soil + SMAC atmosphere to give TOC *and* TOA reflectance at a
# real sensor's own bands), compute indices, save diagnostic figures.
#
# Unlike simulate_RTM() (fourSAIL/foursail2/INFORM), SPART() outputs directly
# at a chosen sensor's bands -- there's no separate "native then convolve"
# step, so "convolution" here means running SPART() once per sensor instead.
# Always 100 simulations. Feeds 2-inversion_ML.R and 3-inversion_DL.R next.
# ==============================================================================

rm(list = ls())

if (!requireNamespace("ToolsRTM", quietly = TRUE)) {
  remotes::install_gitlab("caminoccg/toolsrtm")
}
library(ToolsRTM)
library(ggplot2)

## ----------------------------------------------------------------------------
## User-configurable settings
## ----------------------------------------------------------------------------

n.samples  <- 100                 # always 100 for this course pipeline
leaf.model <- "PROSPECT-PRO"
soil.source <- "marmit"           # "bsm" (SPART's own built-in soil model) or "marmit"
seed <- 1
sensors <- list(Sentinel2A = ToolsRTM::Sentinel2A.MSI, MODIS = ToolsRTM::TerraAqua.MODIS)

correlate.traits <- list(Car = list(with = "Cab", scale = 1/4, r = 0.8))

out_dir <- "outs/ForSPART"
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
pdf(file.path(out_dir, "Rplots.pdf"))  # captures any stray plot() call, see ForPROSAIL/3-simulate_LUT.R for why

cat(sprintf("=== Simulating %d SPART spectra (%s + fourSAIL, soil = %s) ===\n",
            n.samples, leaf.model, soil.source))

## ----------------------------------------------------------------------------
## 1. Build the LUT
## ----------------------------------------------------------------------------

LUT <- as.data.frame(getLUT(inputs = ToolsRTM::inputsSPART, nLUT = n.samples, setseed = seed))

for (trait in names(correlate.traits)) {
  spec <- correlate.traits[[trait]]
  scale_ <- if (is.null(spec$scale)) 1 else spec$scale
  LUT[[trait]] <- correlatedValue(x = LUT[[spec$with]] * scale_, r = spec$r)
  cat(sprintf("Correlation: %s ~ %s (target r=%.2f, realized r=%.2f)\n",
              trait, spec$with, spec$r, cor(LUT[[trait]], LUT[[spec$with]])))
}

# inputsSPART only ships PROSPECT-PRO/-D columns -- add what other leaf
# models need too (same requirement as foursail()/SPART() itself).
LUT$Cs <- 0; LUT$fqe <- 0.01; LUT$Cx <- 0
LUT$cell.d <- 40; LUT$inter.c <- 0.045; LUT$baseline.abs <- 0.0006
LUT$leaf.thick <- 1.6; LUT$albino.abs <- 0; LUT$lign.cell <- 2; LUT$Nitrogen <- 1

key_traits <- intersect(c("Cab", "Car", "Anth", "LAI", "EWT", "LMA", "LIDFa"), names(LUT))

hist_data <- stack(LUT[, key_traits])
colnames(hist_data) <- c("value", "trait")
hist_plot <- ggplot(hist_data, aes(x = value)) +
  geom_histogram(bins = 20, fill = "steelblue", color = "white") +
  facet_wrap(~trait, scales = "free") +
  theme_bw() +
  labs(x = NULL, y = "count", title = sprintf("Sampled trait distributions (n=%d)", n.samples)) +
  theme(plot.title = element_text(hjust = 0.5, size = 12))
print(hist_plot)
ggsave(file.path(out_dir, "0-trait_histograms.png"), plot = hist_plot, width = 18, height = 12, dpi = 300, units = "cm")

corr_mat <- cor(LUT[, key_traits])
png(file.path(out_dir, "0-trait_correlations.png"), width = 1400, height = 1400, res = 200)
corrplot::corrplot(corr_mat, method = "color", type = "upper",
                    addCoef.col = "black", tl.col = "black", tl.srt = 45, number.cex = 0.8,
                    title = "Sampled trait correlations", mar = c(0, 0, 2, 0))
dev.off()
corrplot::corrplot(corr_mat, method = "color", type = "upper",
                    addCoef.col = "black", tl.col = "black", tl.srt = 45, number.cex = 0.8,
                    title = "Sampled trait correlations", mar = c(0, 0, 2, 0))

## ----------------------------------------------------------------------------
## 2. Soil spectrum (SPART wants 400-2400nm, 1nm step, length 2001)
## ----------------------------------------------------------------------------

wl_soil <- 400:2400
if (soil.source == "marmit") {
  soil <- get.marmit.rsoil(database = "Bablet_2016", id = 1, L = 0.05, eps = 0.3, wl.out = wl_soil)
  rsoil <- soil$rsoil.wet
  cat(sprintf("Soil: MARMIT (Bablet_2016 id=1, SMC = %.1f%%)\n", soil$SMC))
} else {
  rsoil <- NULL  # SPART's own built-in BSM soil model
  cat("Soil: SPART's built-in BSM model\n")
}

## ----------------------------------------------------------------------------
## 3. Run SPART for every LUT row, once per sensor
## ----------------------------------------------------------------------------

datasets <- list()
for (sensor_name in names(sensors)) {
  cat(sprintf("Running SPART for sensor = %s ... ", sensor_name))
  sensor_obj <- sensors[[sensor_name]]

  sim_list <- lapply(seq_len(n.samples), function(i) {
    tryCatch(
      suppressMessages(suppressWarnings(SPART(
        inputLUT = LUT[i, ], CanopyModel = "fourSAIL", LeafModel = leaf.model,
        sensor.i = sensor_obj, rsoil = rsoil, get.plots = FALSE
      ))),
      error = function(e) NULL
    )
  })
  ok <- !sapply(sim_list, is.null)
  cat(sprintf("%d/%d rows OK\n", sum(ok), n.samples))

  toc <- t(sapply(sim_list[ok], function(s) s$output$rfl.toc.BRDF))
  toa <- t(sapply(sim_list[ok], function(s) s$output$rfl.toa))
  band_wl <- sim_list[[which(ok)[1]]]$output$wave
  colnames(toc) <- paste0("R.", round(band_wl))
  colnames(toa) <- paste0("R.", round(band_wl))

  indices_toc <- tryCatch(
    suppressMessages(getIndices(as.data.frame(toc), pattern.rfl = "R.", spectral.domain = "VNIR-SWIR")),
    error = function(e) { cat("  indices failed:", conditionMessage(e), "\n"); as.data.frame(toc) }
  )
  indices_toc <- indices_toc[, colSums(is.na(indices_toc)) == 0, drop = FALSE]

  dataset <- cbind(LUT[ok, ], indices_toc)
  dataset <- dataset[, !duplicated(names(dataset))]

  datasets[[sensor_name]] <- list(toc = toc, toa = toa, wl = band_wl, dataset = dataset)
}

## ----------------------------------------------------------------------------
## 4. Diagnostic plot: TOC vs TOA reflectance (Sentinel-2A, by Cab bin)
## ----------------------------------------------------------------------------

ref <- datasets[["Sentinel2A"]]
LUT_ok <- LUT[seq_len(nrow(ref$toc)), ]  # rows are already filtered to the OK subset for this sensor above

plot.compare <- rbind(
  data.frame(wave = rep(ref$wl, each = nrow(ref$toc)), RFL = as.vector(ref$toc), Type = "TOC (canopy)"),
  data.frame(wave = rep(ref$wl, each = nrow(ref$toa)), RFL = as.vector(ref$toa), Type = "TOA (satellite)")
)
toc_toa_plot <- ggplot(plot.compare, aes(x = wave, y = RFL, color = Type)) +
  stat_summary(fun = mean, geom = "line", linewidth = 0.8) +
  stat_summary(fun = mean, geom = "point", size = 2) +
  theme_bw() +
  labs(x = "Wavelength (nm)", y = "Reflectance",
       title = sprintf("SPART: mean TOC vs TOA reflectance, Sentinel-2A bands (n=%d)", nrow(ref$toc))) +
  theme(plot.title = element_text(hjust = 0.5, size = 12))
print(toc_toa_plot)
ggsave(file.path(out_dir, "1-TOC_vs_TOA_reflectance.png"), plot = toc_toa_plot, width = 14, height = 9, dpi = 300, units = "cm")

## ----------------------------------------------------------------------------
## 5. Diagnostic plot: example TOC spectra at Cab's 10th/50th/90th percentile
## ----------------------------------------------------------------------------

pctl_probs <- c(0.1, 0.5, 0.9)
pctl_vals <- quantile(LUT_ok$Cab, probs = pctl_probs)
pctl_rows <- sapply(pctl_vals, function(v) which.min(abs(LUT_ok$Cab - v)))

pctl_long <- do.call(rbind, lapply(seq_along(pctl_rows), function(k) {
  data.frame(wave = ref$wl, RFL = as.numeric(ref$toc[pctl_rows[k], ]),
             Percentile = sprintf("P%d (Cab=%.1f)", pctl_probs[k] * 100, LUT_ok$Cab[pctl_rows[k]]))
}))
pctl_plot <- ggplot(pctl_long, aes(x = wave, y = RFL, color = Percentile)) +
  geom_line(linewidth = 0.8) + geom_point(size = 2) +
  theme_bw() +
  labs(x = "Wavelength (nm)", y = "Reflectance (TOC, Sentinel-2A bands)",
       title = "Example SPART TOC spectra at Cab's 10th/50th/90th percentile") +
  theme(plot.title = element_text(hjust = 0.5, size = 12))
print(pctl_plot)
ggsave(file.path(out_dir, "2-example_spectra_by_percentile.png"), plot = pctl_plot, width = 14, height = 9, dpi = 300, units = "cm")

## ----------------------------------------------------------------------------
## 6. Save
## ----------------------------------------------------------------------------

saveRDS(list(LUT = LUT, datasets = list(native = datasets$Sentinel2A$dataset, se2a = datasets$Sentinel2A$dataset,
                                         modis = datasets$MODIS$dataset),
             leaf.model = leaf.model, soil.source = soil.source, n.samples = n.samples, seed = seed),
        file.path(out_dir, "1-datasets.rds"))

dev.off()

cat("\nSaved to '", out_dir, "/1-datasets.rds' -- read back with readRDS() in 2-inversion_ML.R / ",
    "3-inversion_DL.R ($datasets$native / $se2a / $modis are ready for inversion).\n", sep = "")

File: Scripts/Python/ForSPART_Python/1_simulate_lut.py
"""Simulate a PROSPECT-PRO + fourSAIL SPART LUT: soil + canopy + atmosphere
radiative transfer, giving TOC *and* TOA reflectance at a real sensor's
own bands directly (no separate "native then convolve" step).

Python equivalent of Scripts/R/ForSPART/1-simulate_LUT.R. Unlike
foursail()/foursail2()/inform(), spart_toc()/spart_toa() output straight
at sensor resolution -- "convolution" here means calling spart_toa() once
per sensor instead of a separate resampling step. Atmospheric conditions
(Pa, aot550, uo3, uh2o) use standard clear-sky values (not derived from
altitude like ToolsRTM::get.smac()'s own R fallback) -- a documented
simplification, see the note below.

Feeds 2_inversion_ml.py and 3_inversion_dl.py next.
"""
from pathlib import Path

import numpy as np
import pandas as pd

from toolsrtm import spart_toc, spart_toa, get_indices
from toolsrtm.smac import sentinel2a_msi

OUT_DIR = Path(__file__).resolve().parents[3] / "outs" / "Python" / "ForSPART"
OUT_DIR.mkdir(parents=True, exist_ok=True)

N_SAMPLES = 100
WAVELENGTHS = np.arange(400, 2401)  # spart_toc's own native domain (Fluspect-consistent)

# Standard clear-sky atmosphere -- spart_toa() needs these in inputLUT with
# no built-in default (unlike R's get.smac(), which derives Pa from
# altitude when missing); using fixed typical values is a documented
# simplification of this course pipeline, not a physical measurement.
ATMOSPHERE = dict(Pa=1013.25, aot550=0.2, uo3=0.35, uh2o=2.0)


def sample_lut(n: int, rng: np.random.Generator) -> pd.DataFrame:
    cab = rng.uniform(10, 70, n)
    lut = pd.DataFrame({
        "N": rng.uniform(1.2, 2.5, n),
        "Cab": cab,
        "Car": (cab / 4 + rng.normal(0, 2, n)).clip(min=0.5),
        "Anth": rng.uniform(0, 2, n),
        "Cbrown": rng.uniform(0, 0.3, n),
        "EWT": rng.uniform(0.005, 0.03, n),
        "LMA": rng.uniform(0.003, 0.015, n),
        "alpha": np.full(n, 40.0),
        "Prot": np.full(n, 0.0), "CBC": np.full(n, 0.0),
        "LIDFa": rng.uniform(-0.5, 0.5, n),
        "LIDFb": rng.uniform(-0.3, 0.3, n),
        "TypeLidf": np.full(n, 1, dtype=int),
        "LAI": rng.uniform(0.5, 7.0, n),
        "hspot": np.full(n, 0.01),
        "tts": np.full(n, 30.0),
        "tto": np.full(n, 0.0),
        "psi": np.full(n, 0.0),
        "Pa": np.full(n, ATMOSPHERE["Pa"]), "aot550": np.full(n, ATMOSPHERE["aot550"]),
        "uo3": np.full(n, ATMOSPHERE["uo3"]), "uh2o": np.full(n, ATMOSPHERE["uh2o"]),
    })
    return lut


def main():
    rng = np.random.default_rng(42)
    lut = sample_lut(N_SAMPLES, rng)

    # ---- TOC reflectance (BSM soil, no atmosphere) ----
    toc = np.empty((N_SAMPLES, len(WAVELENGTHS)))
    for i, row in lut.iterrows():
        toc[i] = spart_toc(row.to_dict(), leaf_model="PROSPECT-PRO", SMC=25, BSMBrightness=0.5)
    print(f"Simulated {N_SAMPLES} SPART TOC spectra.")
    indices_toc = get_indices(WAVELENGTHS, toc, spectral_domain="VNIR")
    indices_toc.update(get_indices(WAVELENGTHS, toc, spectral_domain="SWIR"))
    dataset_toc = pd.concat([lut.reset_index(drop=True), pd.DataFrame(indices_toc)], axis=1)
    dataset_toc.to_csv(OUT_DIR / "1_dataset_toc.csv", index=False)
    print(f"TOC: {dataset_toc.shape[1]} columns (traits + reflectance-derived indices)")

    # ---- TOA reflectance at Sentinel-2A's own bands directly ----
    sensor = sentinel2a_msi()
    toa_bands = None
    band_wl = None
    for i, row in lut.iterrows():
        res = spart_toa(row.to_dict(), sensor=sensor, leaf_model="PROSPECT-PRO", SMC=25, BSMBrightness=0.5)
        if toa_bands is None:
            band_wl = np.asarray(res.wl_smac)
            toa_bands = np.empty((N_SAMPLES, len(band_wl)))
        toa_bands[i] = res.rfl_toa
    print(f"Simulated {N_SAMPLES} SPART TOA (Sentinel-2A) spectra, {toa_bands.shape[1]} bands.")
    indices_toa = get_indices(band_wl, toa_bands, spectral_domain="VNIR-SWIR")
    dataset_toa = pd.concat([lut.reset_index(drop=True), pd.DataFrame(indices_toa)], axis=1)
    dataset_toa.to_csv(OUT_DIR / "1_dataset_toa_sentinel2a.csv", index=False)
    print(f"TOA: {dataset_toa.shape[1]} columns")

    print(f"\nSaved -> {OUT_DIR}")


if __name__ == "__main__":
    main()
